### ISSUE TITLE
Question: what are the best practices that you recommend to debug a carthage framework?

### ISSUE THREAD

**Issue by TofPlay** [Carthage/Carthage#1326]
Hi,
I did not see relevant information in the file README.md and I wish to know what are the best practices that you recommend to develop and debug a framework before it release as a carthage framework.
Do I have to include my framework as a sub-project on my application?
Is there a configuration to set up in the project that include this sub-project?

**Comment by mdiep** [comment]
I would typically use `--use-submodules`, include my framework as a sub-project, and develop that way.

Just be aware that if you run a Carthage command, the changes to your dependency could be lost. So commit and push frequently!

**Comment by TofPlay** [comment]
Hi Matt. Thanks for your reply 😊
I have some questions... 😉

**Question 1**: When I use `--use-submodules` all dependencies are declared in the `.gitmodules`. But I just need submodules for my own frameworks not all frameworks. Can I specify that I want only my frameworks as a submodule?

**Question 2**: On the project section `General > Linked Frameworks and Libraries` what framework I must take? The one in `Carthage/Build/iOS/MyFramework.framework` or in `MyFramework.xcodeproj/Products/MyFramework.framework` (in Xcode) ?

**Question 3**: On the project section `Build Phases > Run Script` do I have to let `/usr/local/bin/carthage copy-frameworks` and in section `Input Files` all frameworks used by my project?

**Comment by mdiep** [comment]
> Question 1: When I use --use-submodules all dependencies are declared in the .gitmodules. But I just need submodules for my own frameworks not all frameworks. Can I specify that I want only my frameworks as submodules?

No.

> Question 2: On the project section General > Linked Frameworks and Libraries what framework I must take? The one in Carthage/Build/iOS/MyFramework.framework or in Xcode MyApplication.xcworkspace > MyFramework.xcodeproj > Products > MyFramework.framework?

The latter.

> Question 3: On the project section Build Phases > Run Script do I have to let /usr/local/bin/carthage copy-frameworks and in section Input Files all frameworks used by my project?

`copy-frameworks` is only needed for frameworks that are _built_ with Carthage.

**Comment by TofPlay** [comment]
Hi Matt
Everything seems to work on my current folder. But I have some issue when I try to make a `git clone -b <my branch> <myproject>` to test a new installation from scratch.
When I execute the command `carthage update --use-submodules  --platform iOS` I have this output:
```
$ carthage update --use-submodules --platform iOS
*** Fetching Reachability.swift
*** Fetching realm-cocoa
*** Fetching algoliasearch-client-swift
*** Fetching stripe-ios
*** Fetching AlamofireObjectMapper
*** Fetching ObjectMapper
*** Fetching Alamofire
*** Fetching PureLayout
Parse error: expected submodule commit SHA in output of task (ls-tree -z HEAD Carthage/Checkouts/Alamofire) but encountered:
```
When I check on `Carthage/Checkouts` is empty
What I missing?

**Comment by TofPlay** [comment]
Ok I found it
For Alamofire I was obliged to run those commands:
```bash
$ git submodule add -f https://github.com/Alamofire/Alamofire.git Carthage/Checkouts/Alamofire
$ cd Carthage/Checkouts/Alamofire
$ git checkout tags/3.4.0
```
I had to do the same for other frameworks
My mistake was to believe that with only Cartfile I could restore a functional environment with command
`carthage update --use-submodules --platform iOS` 
We must also commit submodule in `Carthage/Checkouts` and don't forget to clone the main repository with `--recursive` option.

**Suggestion**: Carthage can execute the command `git submodule` to check if all sub-modules are properly installed. If not execute the commands above.

**Comment by TofPlay** [comment]
It's not finish yet 😞
When I try to upload my project on App Store I have this warning
![uplaod to appstore](https://cloud.githubusercontent.com/assets/1082222/15778142/8d547bfc-2994-11e6-9d24-8be845335b51.png)

Any ideas?

**Comment by ikesyo** [comment]
That should not be related to Carthage, that would be an issue on iTC side. See also https://github.com/CocoaPods/CocoaPods/issues/5453.

**Comment by TofPlay** [comment]
@ikesyo Ok in this case just wait and see 🙂  If you have any update can you post a message here? I will do the same.

**Comment by pcantrell** [comment]
> I would typically use --use-submodules, include my framework as a sub-project, and develop that way.

IMO, `use-submodules` should be an option in `Cartfile`, not the command line, because this:

> Just be aware that if you run a Carthage command, the changes to your dependency could be lost. So commit and push frequently!

…is terrible.

Better still, would you be open to supporting a `Cartfile`-relative local path instead of a Github repo? This would have several advantages:

- It would greatly ease local development of a framework.
- Carthage could still support submodules, but get out of the business of managing them and mucking with a project’s git config.
- This would better support frameworks including an example project that uses the framework.

**Comment by mdiep** [comment]
> IMO, use-submodules should be an option in Cartfile, not the command line

#426 is tracking this idea.

> Better still, would you be open to supporting a Cartfile-relative local path instead of a Github repo?

Carthage [already supports this](https://github.com/Carthage/Carthage/blob/master/Documentation/Artifacts.md#example-cartfile).

**Comment by pcantrell** [comment]
>> Better still, would you be open to supporting a Cartfile-relative local path instead of a Github repo?
>
> Carthage already supports this.

As far as I can tell, it only works with absolute paths. Carthage appears to resolve relative paths against its internal cache instead of the `Cartfile`, which renders it useless for the cases I mentioned above:

```
git "file:../.." "master"
```
```
$ carthage update
*** Cloning ..
Failed to check out repository into /Users/paul/Library/Caches/org.carthage.CarthageKit/dependencies/..: No object named "master" exists
```

Same happens with `"../.."`, `"file:/../.."` and `"file://../.."`. Am I being dense and missing something?

**Comment by mdiep** [comment]
A better solution is to just drag the other project into your Xcode project/workspace and build it that way. Just take it out of your Cartfile completely or don't use the copy from your Cartfile while you're developing.

**Comment by pcantrell** [comment]
> A better solution is to just drag the other project into your Xcode project/workspace and build it that way. Just take it out of your Cartfile completely or don't use the copy from your Cartfile while you're developing.

Sure. (Workspace only, right?) But then you’re not testing the framework exactly as Carthage builds it, which has bit me in the past.

**Comment by TofPlay** [comment]
Hi @pcantrell 

> …is terrible.

Yes it's terrible! Having to manage Carthage and git submodule for the same project is not simple. For having tested It's easy to make mistakes. It would have been better if Carthage completely took over git submodules or proposed it's own solution. 
Maybe we'll have something simpler in future versions of Carthage 😉

> Workspace only, right?

Not only on the workspace. You can drag your framework project from `/<your project>/Carthage/Checkouts/<your framework>` under your application project as sub-project. You have to declare your framework with git submodule too. And you have to really be very careful when you manage commits of the main project and those from your framework see as a submodule.
It's what I have done for my framework. As I said above this is not simple.

**Comment by TofPlay** [comment]
I worked around about how I can create and maintain my own frameworks (compatible with Carthage). At the first time the option `--use-submodules` was for me a good thing but in practice is not really simple. Define my framework as a git submodule is not the best but it's ok. But for sure I don't really want submodules for all others frameworks manage by Carthage.  After several trying I declare my framework as submodules and embedded framework in my project. For the others frameworks I use Carthage as usual (without `--use-submodules` ). Because my framework are not manage by Carthage on the main project I create a Cartfile with all dependencies of my frameworks. This way I can maintain my frameworks from the main project. And because each framework has its own repository I can use it as carthage framework on other projects.
I would have preferred all do from Carthage but currently it is not possible. Carthage do not differentiate between the frameworks are used in a project where we just need the binary version and those we developed ourselves and that we must maintain.
Carthage team has already done an amazing job. I am sure they will propose an elegant solution to handle this case. Definitely better than the one I currently use 😋
